snuba/consumer.py

```diff
@@ -110,3 +134,209 @@ def flush_batch(self, batch: Sequence[ProcessedMessage]):
                     )
 
             self.producer.flush()
+
+
+class InsertBatchWriter(ProcessingStep[InsertBatch]):
+    def __init__(self, writer: BatchWriter[WriterTableRow]) -> None:
+        self.__writer = writer
+
+        self.__messages: MutableSequence[Message[InsertBatch]] = []
+        self.__closed = False
+
+    def poll(self) -> None:
+        pass
+
+    def submit(self, message: Message[InsertBatch]) -> None:
+        assert not self.__closed
+
+        self.__messages.append(message)
+
+    def close(self) -> None:
+        self.__closed = True
+
+        if not self.__messages:
+            return
+
+        self.__writer.write(
+            itertools.chain.from_iterable(
+                message.payload.rows for message in self.__messages
+            )
+        )
+
+    def join(self, timeout: Optional[float] = None) -> None:
+        pass
+
+
+class ReplacementBatchWriter(ProcessingStep[ReplacementBatch]):
+    def __init__(self, producer: ConfluentKafkaProducer, topic: Topic) -> None:
+        self.__producer = producer
+        self.__topic = topic
+
+        self.__messages: MutableSequence[Message[ReplacementBatch]] = []
+        self.__closed = False
+
+    def poll(self) -> None:
+        pass
+
+    def submit(self, message: Message[ReplacementBatch]) -> None:
+        assert not self.__closed
+
+        self.__messages.append(message)
+
+    def __delivery_callback(self, error, message) -> None:
+        if error is not None:
+            # errors are KafkaError objects and inherit from BaseException
+            raise error
+
+    def close(self) -> None:
+        self.__closed = True
+
+        if not self.__messages:
+            return
+
+        for message in self.__messages:
+            batch = message.payload
+            key = batch.key.encode("utf-8")
+            for value in batch.values:
+                self.__producer.produce(
+                    self.__topic.name,
+                    key=key,
+                    value=rapidjson.dumps(value).encode("utf-8"),
+                    on_delivery=self.__delivery_callback,
+                )
+
+    def join(self, timeout: Optional[float] = None) -> None:
+        args = []
+        if timeout is not None:
+            args.append(timeout)
+
+        self.__producer.flush(*args)
+
+
+class ProcessedMessageBatchWriter(
+    ProcessingStep[Union[None, InsertBatch, ReplacementBatch]]
+):
```

Comment:
Ideally in the future, `ReplacementBatch` would only be in the type signature for storages that require replacements. All of the `if self.__replacement_batch_writer is not None:` checks in this class are super bodgy.